### PR TITLE
Add set qualifier support in bucket policy condition

### PR DESCRIPTION
**Pull request by balamurugana**
Fixes https://github.com/minio/minio/issues/12026

Signed-off-by: Bala.FA <bala.gluster@gmail.com>

### CODE REVIEW

File: bucket/policy/condition/key.go.orig [hunk 1]
```diff
@@ -0,0 +1,322 @@
+// Copyright (c) 2015-2021 MinIO, Inc.
```
Comment (harshavardhana) [inline bucket/policy/condition/key.go.orig:1]: Looks like an unexpected file @balamurugana

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Review by harshavardhana — CHANGES_REQUESTED** [review]
@balamurugana have you tested this with AWS S3 behavior?

**Comment by harshavardhana** [comment]
> @balamurugana have you tested this with AWS S3 behavior?

@balamurugana checking if you saw this message..

**Comment by balamurugana** [comment]
> > @balamurugana have you tested this with AWS S3 behavior?
> 
> @balamurugana checking if you saw this message..

I am unable to test with AWS S3 because none of the policy is accepted respective to set qualifiers.

**Comment by harshavardhana** [comment]
> > > @balamurugana have you tested this with AWS S3 behavior?
> > 
> > 
> > @balamurugana checking if you saw this message..
> 
> I am unable to test with AWS S3 because none of the policy is accepted respective to set qualifiers.

I am not able to understand @balamurugana are you saying this PR doesn't implement what AWS IAM implements?

**Comment by balamurugana** [comment]
@harshavardhana 
Below are the test results for `ForAllValues` and `ForAnyValue` with AWS S3. Please refer respective unit tests for respective similar behavior in the implementation.

```bash
[bala@localhost mc]$ cat /tmp/policy.json && ./mc policy set-json /tmp/policy.json s3/bala-test-bucket1
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": "s3:ListBucket",
            "Condition": {
                "ForAllValues:StringEquals": {
                    "s3:prefix": [
                        "1",
			"2"
                    ]
                }
            },
            "Effect": "Allow",
            "Principal": {
                "AWS": "*"
            },
            "Resource": "arn:aws:s3:::bala-test-bucket1",
            "Sid": ""
        }
    ]
}
Access permission for `s3/bala-test-bucket1` is set from `/tmp/policy.json`

[bala@localhost mc]$ curl 'https://s3.amazonaws.com/bala-test-bucket1/?prefix=1'
<?xml version="1.0" encoding="UTF-8"?>
<ListBucketResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Name>bala-test-bucket1</Name><Prefix>1</Prefix><Marker></Marker><MaxKeys>1000</MaxKeys><IsTruncated>false</IsTruncated><Contents><Key>123</Key><LastModified>2021-06-23T04:21:58.000Z</LastModified><ETag>&quot;d41d8cd98f00b204e9800998ecf8427e&quot;</ETag><Size>0</Size><StorageClass>STANDARD</StorageClass></Contents></ListBucketResult>

[bala@localhost mc]$ curl 'https://s3.amazonaws.com/bala-test-bucket1/?prefix=2'
<?xml version="1.0" encoding="UTF-8"?>
<ListBucketResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Name>bala-test-bucket1</Name><Prefix>2</Prefix><Marker></Marker><MaxKeys>1000</MaxKeys><IsTruncated>false</IsTruncated><Contents><Key>234</Key><LastModified>2021-06-23T04:22:10.000Z</LastModified><ETag>&quot;d41d8cd98f00b204e9800998ecf8427e&quot;</ETag><Size>0</Size><StorageClass>STANDARD</StorageClass></Contents></ListBucketResult>

[bala@localhost mc]$ curl 'https://s3.amazonaws.com/bala-test-bucket1/?prefix=3'
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>Access Denied</Message><RequestId>00TPP71MDDXF58P9</RequestId><HostId>VbuZBBOD+hdjnvn4q4ZXRBDUIuDPs661QOuQtOOjBbQ1tevV2R38fvgXwyaEqM09Ih8oUrkCWaQ=</HostId></Error>

[bala@localhost mc]$ curl 'https://s3.amazonaws.com/bala-test-bucket1/?delimiter=a'
<?xml version="1.0" encoding="UTF-8"?>
<ListBucketResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Name>bala-test-bucket1</Name><Prefix></Prefix><Marker></Marker><MaxKeys>1000</MaxKeys><Delimiter>a</Delimiter><IsTruncated>false</IsTruncated><Contents><Key>123</Key><LastModified>2021-06-23T04:21:58.000Z</LastModified><ETag>&quot;d41d8cd98f00b204e9800998ecf8427e&quot;</ETag><Size>0</Size><StorageClass>STANDARD</StorageClass></Contents><Contents><Key>234</Key><LastModified>2021-06-23T04:22:10.000Z</LastModified><ETag>&quot;d41d8cd98f00b204e9800998ecf8427e&quot;</ETag><Size>0</Size><StorageClass>STANDARD</StorageClass></Contents><Contents><Key>345</Key><LastModified>2021-06-23T04:22:21.000Z</LastModified><ETag>&quot;d41d8cd98f00b204e9800998ecf8427e&quot;</ETag><Size>0</Size><StorageClass>STANDARD</StorageClass></Contents></ListBucketResult>

[bala@localhost mc]$ curl 'https://s3.amazonaws.com/bala-test-bucket1/'
<?xml version="1.0" encoding="UTF-8"?>
<ListBucketResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Name>bala-test-bucket1</Name><Prefix></Prefix><Marker></Marker><MaxKeys>1000</MaxKeys><IsTruncated>false</IsTruncated><Contents><Key>123</Key><LastModified>2021-06-23T04:21:58.000Z</LastModified><ETag>&quot;d41d8cd98f00b204e9800998ecf8427e&quot;</ETag><Size>0</Size><StorageClass>STANDARD</StorageClass></Contents><Contents><Key>234</Key><LastModified>2021-06-23T04:22:10.000Z</LastModified><ETag>&quot;d41d8cd98f00b204e9800998ecf8427e&quot;</ETag><Size>0</Size><StorageClass>STANDARD</StorageClass></Contents><Contents><Key>345</Key><LastModified>2021-06-23T04:22:21.000Z</LastModified><ETag>&quot;d41d8cd98f00b204e9800998ecf8427e&quot;</ETag><Size>0</Size><StorageClass>STANDARD</StorageClass></Contents></ListBucketResult>
```

```bash
[bala@localhost mc]$ cat /tmp/policy.json && ./mc policy set-json /tmp/policy.json s3/bala-test-bucket1
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": "s3:ListBucket",
            "Condition": {
                "ForAnyValue:StringEquals": {
                    "s3:prefix": [
                        "1",
			"2"
                    ]
                }
            },
            "Effect": "Allow",
            "Principal": {
                "AWS": "*"
            },
            "Resource": "arn:aws:s3:::bala-test-bucket1",
            "Sid": ""
        }
    ]
}
Access permission for `s3/bala-test-bucket1` is set from `/tmp/policy.json`

[bala@localhost mc]$ curl 'https://s3.amazonaws.com/bala-test-bucket1/?prefix=1'
<?xml version="1.0" encoding="UTF-8"?>
<ListBucketResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Name>bala-test-bucket1</Name><Prefix>1</Prefix><Marker></Marker><MaxKeys>1000</MaxKeys><IsTruncated>false</IsTruncated><Contents><Key>123</Key><LastModified>2021-06-23T04:21:58.000Z</LastModified><ETag>&quot;d41d8cd98f00b204e9800998ecf8427e&quot;</ETag><Size>0</Size><StorageClass>STANDARD</StorageClass></Contents></ListBucketResult>

[bala@localhost mc]$ curl 'https://s3.amazonaws.com/bala-test-bucket1/?prefix=2'
<?xml version="1.0" encoding="UTF-8"?>
<ListBucketResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Name>bala-test-bucket1</Name><Prefix>2</Prefix><Marker></Marker><MaxKeys>1000</MaxKeys><IsTruncated>false</IsTruncated><Contents><Key>234</Key><LastModified>2021-06-23T04:22:10.000Z</LastModified><ETag>&quot;d41d8cd98f00b204e9800998ecf8427e&quot;</ETag><Size>0</Size><StorageClass>STANDARD</StorageClass></Contents></ListBucketResult>

[bala@localhost mc]$ curl 'https://s3.amazonaws.com/bala-test-bucket1/?prefix=3'
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>Access Denied</Message><RequestId>NRCS997FTBD4YT03</RequestId><HostId>nVDVB69XcMB9mMKcWdad6cHHkJpryVdbWuV6BzqWZ31hLqipwPG72BxdcOs2IIbophkScmk486k=</HostId></Error>

[bala@localhost mc]$ curl 'https://s3.amazonaws.com/bala-test-bucket1/?delimiter=a'
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>Access Denied</Message><RequestId>6VN6BN6XF8EKQW7D</RequestId><HostId>eg9s3vryZYVEH9VgwUxlpPLPyJ7RIdGJ6mfHbVwcdkboXEYRpqsEJQObUpFJ1XRBqAy0aFB9WD0=</HostId></Error>

[bala@localhost mc]$ curl 'https://s3.amazonaws.com/bala-test-bucket1/'
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>Access Denied</Message><RequestId>BH14WSABDNTCBMWC</RequestId><HostId>c9ypqCal1nwYV8n9WZg7uBtxLJMAf+6HKOy9n4SWXMEkp6uoz7UktvVtrV+VgVQaJwraJmZuEow=</HostId></Error>
```

**Review by krisis — APPROVED** [review]

**Review by klauspost — APPROVED** [review]
LGTM

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**